### ISSUE TITLE
node-check-version.js breaks on new NPM installs

### ISSUE THREAD

**Issue by Pomax** [Automattic/wp-calypso#3048]
https://github.com/Automattic/wp-calypso/blob/66126f311243e8627c96140ff1a3dd00b7eb008f/bin/check-node-version#L3 relies on checking a hardcoded location that no longer applies. A better way to do this would be to use
```
semver = require('semver');
```
with the `semver` package in the devDependencies list.

Although the job this is doing can be done entirely without semver, by simply putting an [`engines`](https://docs.npmjs.com/files/package.json#engines) section in the package.json that says which node version is required for this software to run. No need for `check-node-version` with that in place.

**Comment by klimeryk** [comment]
Thanks for bringing this to our attention, @Pomax! I honestly did not know about the `engines` section. But we _are_ [using it](https://github.com/Automattic/wp-calypso/blob/master/package.json#L109-L111). However, reading into it, it does not quite work as you'd expect. It's only checked by `npm` when installing a package - but not its _dependencies_. And since running `make run` only installs the dependencies, this check is omitted. To trigger it you'd have to, for example, `cd ..` and run `npm install wp-calypso` there ([source](http://www.marcusoft.net/2015/03/packagejson-and-engines-and-enginestrict.html)).

About the changed hardcoded path - which version of `npm` are you referring to?

BTW, we are doing it via script like this because we want to check for a "sane" version of npm, before starting to `npm install` anything. /cc @rralian @blowery for confirmation :grin:

**Comment by Pomax** [comment]
the `process.env.NPM_GLOBAL_ROOT + '/npm/node_modules/semver'` location will not exist for locally installed modules. Simply relying on the local version by using `require('semver')` will still be beneficial to many.

**Comment by nylen** [comment]
More discussion about this at https://github.com/Automattic/wp-calypso/pull/2124/files#r54121505.  Summarizing a bit, we want to run this quick check before `npm install` in case it is a waste of time.

> so far four people I work with ran into the semver require call failing due to it not having been installed

Can you provide more details about this?  Is there a version of Node.js / npm that no longer contains semver in the expected location?

**Comment by Pomax** [comment]
Node does not come with semver at all, it's an external package. It might be that some linux distros bundle some "handy" packages for you for free, but if you install Node from the website (for windows/OSX) or through brew (OSX) or the like, you *just* get Node and NPM, nothing else, so everyone who tries to use Calypso on machines with those runs into `bin/check-node-version` erroring out due to an unmet requirement.

This should be relatively easy to take care of with a `preinstall` hook though:
```
"scripts": {
 ...
 "preinstall": "npm install semver && node bin/check-node-version",
  ...
}
```
This will run as part of an `npm install` call, before the "real" install happens, and any termination with exit code != 0 will cause the install run to abort, so `bin/check-node-version` can call `process.exit(1)` if it sees the wrong version.

**Comment by nylen** [comment]
In every Node.js install I have ever seen, npm is included, and npm depends on semver, so semver is installed.  This includes `brew install node` on OS X.

I'm surprised to hear that there are situations where this isn't true, and I'd like more details about how to reproduce that condition.

**Comment by Pomax** [comment]
Then all I can tell you is that these are regular node.js installs on several OSX and Windows machines. We clone the calypso project, we run the make script, things die when we hit `require("semver")` because it can't be found, and an "npm install semver" solves that problem. So it's probably a great idea still to explicitly add it as `preinstall` operation with the `check-node-version` script call

**Comment by klimeryk** [comment]
As seen in the [Makefile](https://github.com/Automattic/wp-calypso/blob/master/Makefile#L68-L79), we search for the `semver` package in different places depending on the OS:
 * on Windows we search in the base `node_modules` folder for the `node` installation
 * on OSX/Linux, we use the path provided by `npm root -g` and go from there

Can you please add an `echo $SEMVER_GLOBAL_PATH` in the [`node-version` recipe](https://github.com/Automattic/wp-calypso/blob/master/Makefile#L98) and report the path that gets constructed on the affected systems?
```
node-version:
	echo $SEMVER_GLOBAL_PATH
	@$(BIN)/check-node-version
```
That's the path where we expect `semver` to be installed as part of `npm`'s dependencies. As @nylen pointed out, [`semver` is a dependency of `npm`](https://github.com/npm/npm/blob/master/package.json#L92) and it _has_ to be installed for `npm` to work - it's not something installed additionally by any other package manager or custom installer. So it's very puzzling that it's missing (or possibly installed somewhere else?).

**Comment by Pomax** [comment]
I can try to, although I'm still going to recommend not "looking for where it lives" but as a hard dependency, bootstrapping it locally with an `npm install semver`. It will only install the semver package, it won't do the full npm install run, and if you hook it up as preinstall, we're all the way there.

**Comment by Pomax** [comment]
As an aside, a quick check on Debian 8 with Node 5.6 shows the same as windows: 
```
pomax@debian:~$ node -v
v5.6.0
pomax@debian:~$ node
> require('semver')
Error: Cannot find module 'semver'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at repl:1:1
    at REPLServer.defaultEval (repl.js:252:27)
    at bound (domain.js:287:14)
    at REPLServer.runBound [as eval] (domain.js:300:12)
    at REPLServer.<anonymous> (repl.js:417:12)
    at emitOne (events.js:95:20)
> 
```

**Comment by klimeryk** [comment]
> As an aside, a quick check on Debian 8 with Node 5.6 shows the same as windows

That's correct - that's why we use an absolute path (the `SEMVER_GLOBAL_PATH` exported from the Makefile) when requiring `semver` in the `check-node-version` - to pull in `semver` from `npm`'s dependencies.

**Comment by Pomax** [comment]
As a cross-platform advocate, I highly recommend making `npm` do that automatically as part of how it handles installs already. No filesystem mining necessary, so no potential to break on edge cases. Especially now: if I tell you what my `SEMVER_GLOBAL_PATH` var says, is that going to be added as "yet another location to check"? Because that would not be the right solution here.

**Comment by Pomax** [comment]
It looks like "spaces in file names" is biting you here; on Windows at least, Node and NPM install themselves in the "Program Files" folder, so the `SEMVER_GLOBAL_PATH` ends up getting mangled to `c/ Files/nodejs//node_modules/npm/node_modules/semver`, which will never resolve. 

Definitely worth asking npm to handle this instead of trying to circumvent it instead.

**Comment by nylen** [comment]
It does seem cleaner to do `npm install semver && node bin/check-node-version`.  If you'd like to submit a PR for that, feel free to ping any of us for review.

I bet this check isn't the only thing that fails when paths have spaces in them, though.  From the [Windows install instructions](https://github.com/Automattic/wp-calypso/blob/master/docs/windows.md):

> In particular, when installing any of the following tools, be sure to use paths that do not have spaces in them.

**Comment by Pomax** [comment]
yeah that's not really an option for people who actually work with windows setups, though, unless you have the money to buy separate licenses for different concurrent setups =(

Spaces have been in paths for a *long* time now... so any steps towards coping with them will be worth it. I'll file this as PR this weekend, when I have time to look at how much of the make process can actually be done by npm scripts.

**Comment by lancewillett** [comment]
Closing as no action in several months — feel free to re-open once a pull request is submitted to change things.

**Comment by nylen** [comment]
The particular problem here was fixed in #4002.  Any further problems with the build process should get their own specific issue.
